### PR TITLE
Use Radix trees for prefix matching.

### PR DESCRIPTION
**Pull request by klizhentas**
Buffer fan out used simple prefix match
in a loop, what resulted in high CPU load
on many connected watchers.

This commit switches to RADIX trees for
prefix matching what reduces CPU load
substantially for 5K+ connected watchers.

### CODE REVIEW

**Comment by klizhentas** [comment]
retest this please

File: lib/backend/buffer.go [hunk 1]
```diff
@@ -252,3 +285,83 @@ func (w *BufferWatcher) Close() error {
...
+			if buffers[i] == w {
+				buffers = append(buffers[:i], buffers[i+1:]...)
+				found = true
+				break
+			}
+		}
+		if len(buffers) == 0 {
+			t.Tree.Delete(prefix)
+		} else if len(buffers) != prevLen {
+			t.Tree.Insert(prefix, buffers)
```
Comment (r0mant) [inline lib/backend/buffer.go]: Is this gonna replace the existing key?

**Review by r0mant — APPROVED** [review]

File: lib/backend/buffer.go (hunk 1 above)
Comment (klizhentas) [reply lib/backend/buffer.go]: yes

File: lib/backend/buffer.go [hunk 2]
```diff
@@ -191,6 +214,16 @@ func (c *CircularBuffer) NewWatcher(ctx context.Context, watch Watch) (Watcher,
 		watch.QueueSize = len(c.events)
 	}
 
+	if len(watch.Prefixes) == 0 {
+		// if watcher has no prefixes, assume it will match anything
+		// starting from the separator (what includes all keys in backend invariant, see Keys function)
+		watch.Prefixes = append(watch.Prefixes, []byte{Separator})
+	} else {
+		// if watcher's prefixes are redundant, keep only shorter prefixes
+		// to avoid doble fan out
```
Comment (russjones) [inline lib/backend/buffer.go]: double

**Review by russjones — APPROVED** [review]

**Pull request merged**